### PR TITLE
Common health server

### PR DESCRIPTION
**Pull request by tombentley**
### Type of change

- Refactoring

### Description

This PR moves the existing `HealthCheckAndMetricsServer` class from `user-operator` to `operator-common` in preparation for it to be used by the [Unidirectional Topic Operator](https://github.com/strimzi/proposals/blob/main/051-unidirectional-topic-operator.md). The actual logic for liveness and readiness is abstracted into `Liveness` and `Readiness` interfaces, instances of which are passed to the `HealthCheckAndMetricsServer` constructor. Both of these and also the `MetricsProvider` may now be null.

I also reduced the visibility of the inner classes, added a bit of logging and improved the coverage of the test.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [x] Make sure all tests pass

### CODE REVIEW

**Review by scholzj — APPROVED** [review]
LGTM. I wonder if `http` is the right package name -> it is not really any generic HTTP stuff but health check and metrics server. But its not ike I have any better ide which would be short 🤔 .

**Comment by tombentley** [comment]
> I wonder if `http` is the right package name

I agree, it's not _quite_ right. Happy to change it we can agree to something better. It's almost `observability`, or `management`, or something, but those suffer from being a bit unspecific.

**Comment by tombentley** [comment]
What about `scrape`?

**Comment by scholzj** [comment]
I don't think anyone without prior knowledge would understand what it means.

File: operator-common/src/main/java/io/strimzi/operator/common/http/Liveness.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+package io.strimzi.operator.common.http;
+
+/**
+ * A liveness check implemented by an application (e.g. an operator) and called by the
+ * {@link HealthCheckAndMetricsServer} when handling a health check request.
+ */
+public interface Liveness {
+
+    /**
+     * Indicates whether the controller is alive or not.
```
Comment (ppatierno) [inline operator-common/src/main/java/io/strimzi/operator/common/http/Liveness.java:14]: isn't "the application" better? You are being generic later about it and then adding an example usage with a controller, or?

File: operator-common/src/main/java/io/strimzi/operator/common/http/Readiness.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+package io.strimzi.operator.common.http;
+
+/**
+ * A readiness check implemented by an application (e.g. an operator) and called by the
+ * {@link HealthCheckAndMetricsServer} when handling a health check request.
+ */
+public interface Readiness {
+
+    /**
+     * Indicates whether the controller is ready.
```
Comment (ppatierno) [inline operator-common/src/main/java/io/strimzi/operator/common/http/Readiness.java:14]: Ditto as above.

**Review by ppatierno — APPROVED** [review]
LGTM. Just left a couple of nits.

File: user-operator/src/main/java/io/strimzi/operator/user/Main.java [hunk 3]
```diff
@@ -99,7 +100,7 @@ public static void main(String[] args) {
         );
 
         // Create the health check and metrics server
-        HealthCheckAndMetricsServer healthCheckAndMetricsServer = new HealthCheckAndMetricsServer(controller, metricsProvider);
+        HealthCheckAndMetricsServer healthCheckAndMetricsServer = new HealthCheckAndMetricsServer(controller, controller, metricsProvider);
```
Comment (fvaleri) [inline user-operator/src/main/java/io/strimzi/operator/user/Main.java:103]: We could create a common interface that extends Liveness, Readiness and MetricsProvider. Then we could use it as an additional one-argument constructor for the HealthCheckAndMetricsServer. WDYT?

**Pull request merged**